File: Scripts/Sound/SoundManager.js
#pragma downcast
//@script RequireComponent(AudioSource)

var soundOn : boolean = true;
var soundVolume : float = 1.0;
var musicVolume : float = 1.0;
var soundNames : String[];
var sounds : AudioClip[];
static var soundMap : Hashtable = new Hashtable();

// We essentially create a static object pool, with a set number of
// AudioSources ('channels'). We cycle through these each time we play
// a sound - this will reduce the change of one sample somehow clobbering
// another, and allows one channel to play at a different pitch without
// disrupting other playing samples.
public var numberOfChannels : int;
private var channels : AudioSource[];
private var channelIndex : int = 0;

// A dedicated background music channel
// (TODO: probably really want a pool of two of these to allow cross fading between two tracks)
private var musicChannel : AudioSource;

function Awake () {
  if (numberOfChannels == null) {
  	numberOfChannels = 8;
  }
  channels = new AudioSource[numberOfChannels];
  for (var i : int = 0; i < soundNames.Length; i++) {
    try {
      soundMap.Add(soundNames[i], sounds[i]);
    } catch (err) {}
  }
  /*
  soundMap = {"Die": sounds[0],
            "BeamIn" : sounds[1],
            "PowerupPickup": sounds[2] 
            };
  */
  
  //musicChannel = transform.Find("MusicChannel").GetComponentInChildren(typeof(AudioSource));
  var musicChannelGO : GameObject = new GameObject();
  musicChannelGO.name = "MusicChannel";
  musicChannelGO.transform.parent = transform;
  musicChannel = musicChannelGO.AddComponent(AudioSource);
  musicChannel.audio.loop = true;
  
  // generate attached gameobjects "Channel1", "Channel2" with AudioSources
  // accessible in the channel array
  for (var ii : int = 0; ii < numberOfChannels; ii++) {
  	var c : GameObject = new GameObject();
  	c.name = "Channel"+ii;
  	c.transform.parent = transform;
  	c.AddComponent(AudioSource);
  	c.audio.loop = false;
  	channels[ii] = c.audio;
  }
          
  DontDestroyOnLoad(transform.gameObject);
}

function Start () {

}

function Update () {

}

function Play(soundname : String, volume : float, pitch : float) {
  if (soundOn) {
    while (channels[channelIndex].isPlaying) {
      incrementChannelIndex();
    }
    channels[channelIndex].pitch = pitch;
    channels[channelIndex].volume = volume;
  	channels[channelIndex].PlayOneShot(soundMap[soundname], soundVolume);
  	incrementChannelIndex();
  }
}

function Play(soundname : String) {
	Play(soundname, soundVolume, 1.0);
}

function Play(soundname : String, volume : float) {
	Play(soundname, volume, 1.0);
}

private function incrementChannelIndex() {
  	if (channelIndex < channels.Length - 1) {
  	   channelIndex++;
  	} else {
  	   channelIndex = 0;
  	}
}

// TODO: function to switch between two music tracks at different speed.
// Use 
// var percent_played : float = GetComponent(AudioSource).timeSamples / GetComponent(AudioSource).clip.samples
// GetComponent(AudioSource).clip.soundMap["FasterMusic"];
// GetComponent(AudioSource).timeSamples = percent_played * GetComponent(AudioSource).clip.samples


function SwitchToMusic(soundname : String) {
  if (soundOn) {
    var percent_played : float = (1.0*musicChannel.timeSamples) / (1.0*musicChannel.clip.samples);
    musicChannel.clip = soundMap[soundname];
    musicChannel.volume = musicVolume;
    musicChannel.loop = true;
  	//musicChannel.Play(percent_played * musicChannel.clip.samples);
  	musicChannel.timeSamples = parseInt(percent_played * musicChannel.clip.samples);
  	musicChannel.Play();
  	Debug.Log(soundname + ": " + percent_played);
  }
}

function PlayMusic(soundname : String) {
  if (soundOn) {
    musicChannel.volume = musicVolume;
    musicChannel.pitch = 1.0;
    musicChannel.clip = soundMap[soundname];
    musicChannel.loop = true;
  	musicChannel.Play();
  }
}

function PlayMusic(soundname : String, volume : float, pitch : float) {
  if (soundOn) {
    musicChannel.volume = Mathf.Min(musicVolume, volume);
    musicChannel.pitch = pitch;
    musicChannel.clip = soundMap[soundname];
    musicChannel.loop = true;
  	musicChannel.Play();
  }
}

function PauseMusic() {
	musicChannel.Pause();
}

function StopMusic() {
	musicChannel.Stop();
}

File: Scripts/Sound/SoundManager.cs
using UnityEngine;
using System.Collections;
using System.Collections.Generic;

public class SoundManager : MonoBehaviour {
    public bool soundOn = true;
    public enum CameraFollowMode {
        None,
        Follow,
        Child
    };
    public CameraFollowMode cameraFollowMode = CameraFollowMode.Follow;
    public float volume = 1.0f;
	public bool ScaleOutputVolume = true;
    public string preferenceName = "sfx_volume";
    public bool initialVolumeFromPreference = true;
    public bool useFilenameAsSoundName = true;
    public List<string> soundNames;
    public AudioClip[] sounds;
    Dictionary<string,AudioClip> soundMap = new Dictionary<string,AudioClip>();
 
    // we keep tally of how many sfx of a particular name are playing simultaneously
    // and don't play more than a certain number at a time (for efficiency doesn't 
    // account for sounds that are manually stopped, assumes they play for the full duration
    // after being triggered)
    public int maxSimlutSfxOfOneType = 4;
    Dictionary<string,int> currentlyPlayingCount = new Dictionary<string,int>();
 
    // We essentially create a static object pool, with a set number of
    // AudioSources ('channels'). We cycle through these each time we play
    // a sound - this will reduce the change of one sample somehow clobbering
    // another, and allows one channel to play at a different pitch without
    // disrupting other playing samples.
    public int numberOfChannels = 16;
    private AudioSource[] channels;
    private int channelIndex = 0;

    // Static singleton property
    public static SoundManager Instance { get; private set; }
 
    public void Awake() {
        /// singleton stuff ///////
        // First we check if there are any other instances conflicting
        if (Instance != null && Instance != this) {
            // If that is the case, we destroy this instance
            Destroy(gameObject);
            return;
        }
 
        // Here we save our singleton instance
        Instance = this;
 
        // Furthermore we make sure that we don't destroy between scenes (this is optional)
        DontDestroyOnLoad(gameObject);
        /////////////////////////////

        channels = new AudioSource[numberOfChannels];
        for (int i = 0; i < sounds.Length; i++) {
            if (useFilenameAsSoundName) {
                soundNames.Add(sounds[i].name);
                soundMap.Add(sounds[i].name, sounds[i]);
            } else {
                soundMap.Add(soundNames[i], sounds[i]);
            }
            currentlyPlayingCount.Add(soundNames[i], 0);
        }
         
        // generate attached gameobjects "Channel1", "Channel2" with AudioSources
        // accessible in the channel array
        for (int ii = 0; ii < numberOfChannels; ii++) {
            GameObject c = new GameObject();
            c.name = "Channel" + ii;
            c.transform.parent = transform;
            AudioSource a = c.AddComponent<AudioSource>();
            a.rolloffMode = AudioRolloffMode.Linear;
            a.panLevel = 0f;
            a.dopplerLevel = 0f;
            c.audio.loop = false;
            channels[ii] = c.audio;
        }
         
        DontDestroyOnLoad(transform.gameObject);
 
        if (cameraFollowMode == CameraFollowMode.Follow) {
            if (Camera.main != null)
                transform.position = Camera.main.transform.position;
        } else if (cameraFollowMode == CameraFollowMode.Child) {
            if (Camera.main != null)
                transform.parent = Camera.main.transform;
        } 
     
        if (initialVolumeFromPreference)
            volume = GetVolumePreference();
    }
 
    void OnLevelWasLoaded(int level) {
        if (cameraFollowMode == CameraFollowMode.Follow) {
            if (Camera.main != null)
                transform.position = Camera.main.transform.position;
        } else if (cameraFollowMode == CameraFollowMode.Child) {
            if (Camera.main != null)
                transform.parent = Camera.main.transform;
        } 
    }
 
    public void Start() {
 
    }
 
    public void Update() {
        if (cameraFollowMode == CameraFollowMode.Follow) {
            if (Camera.main != null)
                transform.position = Camera.main.transform.position;
        }
    }
 
    public float GetVolumePreference() {
        return PlayerPrefs.GetFloat(preferenceName, volume);
    }
 
    public void SaveCurrentVolumePreference() {
        SaveVolumePreference(volume);
    }
 
    public void SaveVolumePreference(float v) {
        PlayerPrefs.SetFloat(preferenceName, v);
        PlayerPrefs.Save();
    }
 
    public void Play(string soundname) {
        Play(soundname, volume, 1.0f);
    }
 
    public void Play(string soundname, float volume, float pitch) {
        if (!soundMap.ContainsKey(soundname)) {
            Debug.LogWarning("SoundManager: Tried to play undefined sound: " + soundname);
            return;
        }
        if (soundOn) {
            if (currentlyPlayingCount[soundname] > maxSimlutSfxOfOneType) {
                return;
            }
            int channelsPlayingCount = 0;
            while(channels[channelIndex].isPlaying) {
                incrementChannelIndex();
                channelsPlayingCount++;
            }
            // skip playing this effect if all channels are in use
            if (channelsPlayingCount >= numberOfChannels) {
                return;
            }
            channels[channelIndex].pitch = pitch;
            channels[channelIndex].volume = volume;
			float v = volume;
			if (ScaleOutputVolume) v = ScaleVolume(volume);
            channels[channelIndex].PlayOneShot(soundMap[soundname], v);
            currentlyPlayingCount[soundname] += 1;
            StartCoroutine("decrementPlayCountInFuture", soundname);
            incrementChannelIndex();
        }
    }
 
    public void Play(string soundname, float volume) {
        Play(soundname, volume, 1.0f);
    }
 
	// TODO: we should consider using this dB scale as an option when porting these changes 
	//       over to unity-bowerbird: http://wiki.unity3d.com/index.php?title=Loudness
	/*
	 *   Quadratic scaling of actual volume used by AudioSource. Approximates the proper exponential.
	 */
	public float ScaleVolume(float v) {
		v = Mathf.Pow(v, 4);
		return Mathf.Clamp01(v);
	}

    private void incrementChannelIndex() {
        if (channelIndex < channels.Length - 1) {
            channelIndex++;
        } else {
            channelIndex = 0;
        }
    }
 
    private IEnumerator decrementPlayCountInFuture(string soundname) {
        if (Time.timeScale > 0) {
            yield return new WaitForSeconds(soundMap[soundname].length);
        }
        if (currentlyPlayingCount[soundname] > 0)
            currentlyPlayingCount[soundname] -= 1;
    }
}
